rfcs/20251203-grouped-gemm-support/README.md

```diff
@@ -0,0 +1,604 @@
+float* weight_data;             // [Expert0_weights | Expert1_weights | ...]
+```
+
+- Offsets and sizes have dynamic size based on active ((!) not total) experts (therefore extra array is needed),
+  that is a bit different from the PyTorch/proposed oneDNN design, but compatible.
+- Current version provides offsets on host,
+  but device-side offsets expected for the future GPU-based routing.
+
+## oneDNN Grouped GEMM Support
+
+The request is to add support for grouped GEMM operation in oneDNN.
+The main target is GPU execution, since MoE workloads are primarily run on GPUs,
+and ultimately we want to reduce number of kernel launches and be able to handle
+the runtime variable dimensions per expert located on device side after routing.
+
+### Proposal #1: Via New Grouped Memory Descriptor (Recommended)
+
+This proposal adds a new grouped memory descriptor that works with the
+existing matmul primitive.
+
+The idea is to represent grouped GEMM data as a "sparse" encoding similar
+to CSR format. Data is stored as concatenated blocks where each expert
+processes a different number of tokens.
+
+Representation is:
+```
+Contiguous memory: [Expert0 | Expert1 | Expert2 | Expert3]
+Block sizes:       [M0 x K  | M1 x K  | M2 x K  | M3 x K ]
+
+Offsets array is for tracking where each expert starts:
+offsets = [0, M0, M0+M1, M0+M1+M2, M0+M1+M2+M3]
+
+Note: Why "sparse"? M_i can be zero if an expert receives no tokens.
+In that case, the next expert starts at the same offset, similar to CSR format row pointers.
+Example: if M1 = 0, then offsets = [0, M0, M0, M0 + M2, M0 + M2 + M3]
+```
+
+The grouped memory descriptor defines:
+- Block count: `num_experts` (known at creation)
+- Total elements: `num_input_tokens x TOP_K x K` (known at creation)
+- Block dimensions: `[Mi x K]` where `Mi` is runtime, `K` is fixed
+- Data types: values and offsets (known at creation)
+- Values buffer: Contiguous data sum(Mi x K) (provided at execution)
+- Offsets buffer: `[0, M0, M0 + M1, ...]` size `num_experts + 1` (at execution)
+
+This representation is analogous to CSR sparse format. CSR uses row
+pointers and column indices to describe sparse structure. The grouped
+descriptor uses offsets to mark block boundaries.
+Empty blocks (experts with zero tokens) are skipped via offsets.
+Structure is known at primitive creation, while runtime dimensions and buffers
+are resolved at execution time using device-side data.
+
+#### Memory Descriptor API
+
+The grouped memory descriptor is created similarly to CSR descriptors:
+
+```cpp
+static memory::desc grouped(
+    const dims &adims,               // Overall tensor dims [total_M, K]
+    data_type adata_type,            // Elements data type
+    dim group_count,                 // Group size or number of experts (all, not just active ones)
+    const dims &grouped_dims,        // Group dimensions (clarification below)
+    data_type offsets_dt = s32,      // Offset buffer dtype (default int32)
```

Comment:
Given that this is an address offset, maybe it would be safer to make it s64 ? (this is comming with dropout attribute extension PR).